File: backend/app/api/v1/routes/activities.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncConnection

from app.api.errors import raise_api_error
from app.api.v1.dependencies import get_current_user_id, get_optional_user_id
from app.api.v1.repositories.activity_repo import (
    cancel_activity,
    create_activity,
    get_activity_or_404,
    join_activity,
    leave_activity,
    list_activities,
    list_participants,
    update_activity,
)
from app.api.v1.schemas import (
    ActivityCreate,
    ActivityDetail,
    ActivityListItem,
    ActivityStatus,
    Page,
    Participant,
)
from app.db import get_connection

router = APIRouter()


@router.get("", response_model=Page[ActivityListItem])
async def list_activities_route(
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    viewer_user_id: Annotated[UUID | None, Depends(get_optional_user_id)],
    city_id: UUID | None = None,
    category: Annotated[str | None, Query(min_length=2, max_length=60)] = None,
    status_filter: Annotated[ActivityStatus | None, Query(alias="status")] = None,
    search: Annotated[str | None, Query(min_length=2, max_length=200)] = None,
    lat: Annotated[float | None, Query(ge=-90, le=90)] = None,
    lon: Annotated[float | None, Query(ge=-180, le=180)] = None,
    radius_m: Annotated[int | None, Query(ge=100, le=100_000)] = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 30,
    cursor: str | None = None,
) -> Page[ActivityListItem]:
    if (lat is None) != (lon is None):
        raise_api_error(status.HTTP_400_BAD_REQUEST, "geo_point_required", "lat and lon must be provided together")

    return await list_activities(
        connection=connection,
        viewer_user_id=viewer_user_id,
        limit=limit,
        cursor=cursor,
        city_id=city_id,
        category=category,
        status_filter=status_filter,
        search=search,
        lat=lat,
        lon=lon,
        radius_m=radius_m,
    )


@router.post("", response_model=ActivityDetail, status_code=status.HTTP_201_CREATED)
async def create_activity_route(
    payload: ActivityCreate,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    current_user_id: Annotated[UUID, Depends(get_current_user_id)],
) -> ActivityDetail:
    return await create_activity(connection, payload, current_user_id)


@router.get("/{activity_id}", response_model=ActivityDetail)
async def get_activity_route(
    activity_id: UUID,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    viewer_user_id: Annotated[UUID | None, Depends(get_optional_user_id)],
) -> ActivityDetail:
    return await get_activity_or_404(connection, activity_id, viewer_user_id)


@router.patch("/{activity_id}", response_model=ActivityDetail)
async def update_activity_route(
    activity_id: UUID,
    payload: ActivityCreate,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    current_user_id: Annotated[UUID, Depends(get_current_user_id)],
) -> ActivityDetail:
    return await update_activity(connection, activity_id, payload, current_user_id)


@router.get("/{activity_id}/participants", response_model=list[Participant])
async def list_participants_route(
    activity_id: UUID,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
) -> list[Participant]:
    await get_activity_or_404(connection, activity_id, None)
    return await list_participants(connection, activity_id)


@router.post("/{activity_id}/join", response_model=ActivityDetail)
async def join_activity_route(
    activity_id: UUID,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    current_user_id: Annotated[UUID, Depends(get_current_user_id)],
) -> ActivityDetail:
    return await join_activity(connection, activity_id, current_user_id)


@router.post("/{activity_id}/leave", response_model=ActivityDetail)
async def leave_activity_route(
    activity_id: UUID,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    current_user_id: Annotated[UUID, Depends(get_current_user_id)],
) -> ActivityDetail:
    return await leave_activity(connection, activity_id, current_user_id)


@router.post("/{activity_id}/cancel", response_model=ActivityDetail)
async def cancel_activity_route(
    activity_id: UUID,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    current_user_id: Annotated[UUID, Depends(get_current_user_id)],
) -> ActivityDetail:
    return await cancel_activity(connection, activity_id, current_user_id)

File: HubNear/src/app/api/activities.ts
import { apiRequest } from "./client";
import type { ActivityCreate, ActivityDetail, ActivityListItem, Page, Participant } from "./types";

export function listActivities(params?: {
  city_id?: string;
  category?: string;
  status?: string;
  search?: string;
  lat?: number;
  lon?: number;
  radius_m?: number;
  limit?: number;
  cursor?: string;
}): Promise<Page<ActivityListItem>> {
  const searchParams = new URLSearchParams();
  if (params?.city_id) searchParams.set("city_id", params.city_id);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.status) searchParams.set("status", params.status);
  if (params?.search) searchParams.set("search", params.search);
  if (params?.lat !== undefined) searchParams.set("lat", String(params.lat));
  if (params?.lon !== undefined) searchParams.set("lon", String(params.lon));
  if (params?.radius_m) searchParams.set("radius_m", String(params.radius_m));
  if (params?.limit) searchParams.set("limit", String(params.limit));
  if (params?.cursor) searchParams.set("cursor", params.cursor);
  const qs = searchParams.toString();
  return apiRequest(`/activities${qs ? `?${qs}` : ""}`);
}

export function getActivity(id: string): Promise<ActivityDetail> {
  return apiRequest(`/activities/${id}`);
}

export function createActivity(data: ActivityCreate): Promise<ActivityDetail> {
  return apiRequest("/activities", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function listParticipants(activityId: string): Promise<Participant[]> {
  return apiRequest(`/activities/${activityId}/participants`);
}

export function joinActivity(activityId: string): Promise<ActivityDetail> {
  return apiRequest(`/activities/${activityId}/join`, { method: "POST" });
}

export function leaveActivity(activityId: string): Promise<ActivityDetail> {
  return apiRequest(`/activities/${activityId}/leave`, { method: "POST" });
}

export function cancelActivity(activityId: string): Promise<ActivityDetail> {
  return apiRequest(`/activities/${activityId}/cancel`, { method: "POST" });
}
